### PR TITLE
Compute, Don't Store, DOCC/SOCC

### PR DESCRIPTION
**Pull request by JonathonMisiewicz**
## Description
#2476 and #2594 indicate a major problem in Psi's SCF code: the simplistic formula `nalphapi = doccpi + soccpi` and `nbetapi = doccpi` is incapable of describing references where there are more beta than alpha orbitals of a given irrep. This occurs in MOM (where we instead use ugly hacks) and in UHF (where we crash). Furthermore, storing all of these irrep quantities is redundant.

This PR remedies the situation by only storing `nalphapi` and `nbetapi`. `doccpi` and `soccpi` can be computed from these when needed. The Py-side API is unchanged `doccpi(), soccpi()`, while the C-side API changes from `doccpi_, soccpi_` to `doccpi(), soccpi()`. While we are doing some extra work to compute docc and socc every time they're needed, the computational cost is negligible in comparison to Fock diagonalizations, integral transforms, and tensor contractions. 

To prevent scope creep, this PR aims solely to change the wavefunction. **This is a major undertaking and should not be done lightly.** A subsequent PR will fix the linked issues (if not fixed by this PR) and add them as test cases once the fix is confirmed.

Obligatory @susilehtola ping.

## Todos
- [x] DOCC and SOCC are now computed rather than stored
- [x] MOM simplified
- [x] Lots of docc/socc replaced with alphapi/betapi where more appropriate
- [x] More auto
- [x] More dimension, fewer raw arrays 

## Checklist
- [x] Passes ctest and pytest (all, except addons)

## Status
- [x] Ready for review
- [x] Ready for merge

### CODE REVIEW

**Comment by JonathonMisiewicz** [comment]
Eco failure is, quite predictably, because it breaks `v2rdm`.

**Comment by JonathonMisiewicz** [comment]
After discussion with Lori, work on getting this PR through will resume after MQM. We also have a plan to deal with plugin incompatibilities.

Feel free to review earlier if you wish.

**Review by loriab — APPROVED** [review]
quite an undertaking, lgtm!

**Review by jturney — APPROVED** [review]

File: psi4/src/psi4/adc/adc.cc [hunk 1]
```diff
@@ -45,53 +45,48 @@ ADCWfn::ADCWfn(SharedWavefunction ref_wfn, Options& options) : Wavefunction(opti
...
 
     std::vector<std::string> irreps_ = molecule_->irrep_labels();
-    aoccpi_ = new int[nirrep_];
-    boccpi_ = new int[nirrep_];
-    avirpi_ = new int[nirrep_];
-    bvirpi_ = new int[nirrep_];
+    aoccpi_ = nalphapi_ - frzcpi_;
+    boccpi_ = nbetapi_ - frzcpi_;
+    avirpi_ = nmopi_ - nalphapi_ - frzvpi_;
+    bvirpi_ = nmopi_ - nbetapi_ - frzvpi_;
```
Comment (susilehtola) [inline psi4/src/psi4/adc/adc.cc:51]: Aesthetic change
```suggestion
    avirpi_ = nmopi_ - frzvpi_ - nalphapi_;
    bvirpi_ = nmopi_ - frzvpi_ - nbetapi_;
```

File: psi4/src/psi4/libmints/wavefunction.h [hunk 2]
```diff
@@ -378,10 +374,22 @@ class PSI_API Wavefunction : public std::enable_shared_from_this<Wavefunction> {
...
+        return docc_vec;
+    }
+    /// Returns the SOCC per irrep array. Not recommended for unrestricted code.
+    const Dimension soccpi() const {
+        std::vector<int> socc_vec;
+        for (int h = 0; h < nalphapi_.n(); h++) {
+            socc_vec.push_back(std::abs(nalphapi_[h] - nbetapi_[h]));
+        }
+        return socc_vec;
+    }
```
Comment (susilehtola) [inline psi4/src/psi4/libmints/wavefunction.h:392]: Should this code issue a warning when problematic cases are encountered? In that case, results of any code using socc and docc are incorrect...

Comment (JonathonMisiewicz) [reply psi4/src/psi4/libmints/wavefunction.h:392]: Problematic meaning `nbeta > nalpha` for at least one irrep, or something else?

File: psi4/src/psi4/adc/adc.cc (hunk 1 above)
Comment (JonathonMisiewicz) [reply psi4/src/psi4/adc/adc.cc:51]: Because there's something wrong with the test suite, I'll make the change only if I need to issue a new commit for some other reason. I agree aesthetically, but it's not worth a delay to the PR getting in. The buggy ADC code is not long for this world anyways. 🙂

File: psi4/src/psi4/libmints/wavefunction.h (hunk 2 above)
Comment (susilehtola) [reply psi4/src/psi4/libmints/wavefunction.h:392]: > Problematic meaning `nbeta > nalpha` for at least one irrep, or something else?

Exactly

Comment (JonathonMisiewicz) [reply psi4/src/psi4/libmints/wavefunction.h:392]: @loriab Sound like a good candidate for the log file?

Comment (loriab) [reply psi4/src/psi4/libmints/wavefunction.h:392]: It'll be the first warning for the log file, but sounds appropriate. Could also repeat in out file if you want.

**Comment by JonathonMisiewicz** [comment]
@susilehtola There was a request among core devs for you to sign off on this PR before we get a third review.

File: psi4/src/psi4/libmints/wavefunction.cc [hunk 3]
```diff
@@ -753,6 +739,28 @@ void Wavefunction::initialize_singletons() {
...
+    return docc_vec;
+}
+/// Returns the SOCC per irrep array. Not recommended for unrestricted code.
+const Dimension Wavefunction::soccpi(bool assume_socc_alpha) const {
+    std::vector<int> socc_vec;
+    for (int h = 0; h < nalphapi_.n(); h++) {
+        socc_vec.push_back(std::abs(nalphapi_[h] - nbetapi_[h]));
+        if (assume_socc_alpha && nbetapi_[h] > nalphapi_[h]) {
+            outfile->Printf("Warning! Irrep has more beta than alpha electrons.\n");
+        }
```
Comment (susilehtola) [inline psi4/src/psi4/libmints/wavefunction.cc:759]: `assume_socc_alpha` is not properly handled here

File: psi4/src/psi4/libmints/wavefunction.cc [hunk 4]
```diff
@@ -804,16 +812,16 @@ void Wavefunction::force_soccpi(const Dimension &soccpi) {
...
     for (int h = 0; h < nirrep_; h++) {
-        if ((soccpi[h] + doccpi_[h]) > nmopi_[h]) {
+        if ((soccpi[h] + doccpi()[h]) > nmopi_[h]) {
             throw PSIEXCEPTION(
                 "Wavefunction::force_soccpi: Number of doubly and singly occupied orbitals in an irrep cannot exceed "
                 "the total number of molecular orbitals.");
         }
-        soccpi_[h] = soccpi[h];
-        nalphapi_[h] = doccpi_[h] + soccpi_[h];
+        nalphapi_[h] = docc[h] + soccpi[h];
```
Comment (susilehtola) [inline psi4/src/psi4/libmints/wavefunction.cc:822]: What's the idea with `docc`, `doccpi()`, and `doccpi_`? Wh yso many variables that assumedly are the same?

File: psi4/src/psi4/libmints/wavefunction.cc (hunk 3 above)
Comment (JonathonMisiewicz) [reply psi4/src/psi4/libmints/wavefunction.cc:759]: How? I don't see an issue. If you assume that all SOCC are alpha _and_ observe that there are more beta than alpha for a given irrep, you have a problem. That is exactly the problem I am trying to catch.

File: psi4/src/psi4/libmints/wavefunction.cc (hunk 4 above)
Comment (JonathonMisiewicz) [reply psi4/src/psi4/libmints/wavefunction.cc:822]: I can change l. 817 to use `docc` instead of `doccpi()`. The point is to avoid recompute.

File: psi4/src/psi4/libmints/wavefunction.cc (hunk 3 above)
Comment (susilehtola) [reply psi4/src/psi4/libmints/wavefunction.cc:759]: My issue is that it's not clear what the code should be doing for `assume_socc_alpha = false`.

Comment (JonathonMisiewicz) [reply psi4/src/psi4/libmints/wavefunction.cc:759]: That we are not assuming all SOCC are alpha does not imply that we are assuming all SOCC are not alpha. If you can find a better argument name, I can change it. `warn_on_beta_socc`?

Comment (JonathonMisiewicz) [reply psi4/src/psi4/libmints/wavefunction.cc:759]: @susilehtola

File: psi4/src/psi4/libmints/wavefunction.cc (hunk 4 above)
Comment (susilehtola) [reply psi4/src/psi4/libmints/wavefunction.cc:822]: What is the difference between `docc`, `doccpi()` and `doccpi_`?

Comment (susilehtola) [reply psi4/src/psi4/libmints/wavefunction.cc:822]: Shouldn't `soccpi_` and `doccpi_` be removed?

File: external/downstream/v2rdm_casscf/CMakeLists.txt [hunk 5]
```diff
@@ -16,7 +16,7 @@ if(${ENABLE_v2rdm_casscf})
 
         ExternalProject_Add(v2rdm_casscf_external
             DEPENDS psi4-core
-            URL https://github.com/loriab/v2rdm_casscf/archive/929d7fc.tar.gz
+            URL https://github.com/loriab/v2rdm_casscf/archive/v2rdm8.tar.gz
```
Comment (susilehtola) [inline external/downstream/v2rdm_casscf/CMakeLists.txt:19]: What's the reason for this change? It does not appear to be related to docc/socc.

Put it in a separate pull.

File: psi4/src/psi4/adc/adc.cc [hunk 6]
```diff
@@ -45,53 +45,48 @@ ADCWfn::ADCWfn(SharedWavefunction ref_wfn, Options& options) : Wavefunction(opti
...
     avire_ = new double[navir];
     bvire_ = new double[nbvir];
 
-    nopen_ = 0;
-    for (int h = 0; h < nirrep_; h++) nopen_ += soccpi_[h];
+    nopen_ = soccpi().sum();
     if (!psio_) {
         throw PSIEXCEPTION("The wavefunction passed in lacks a PSIO object, crashing ADC. See GitHub issue #1851.");
     }
     if (nopen_) throw PSIEXCEPTION("Openshell calculation has not been implemented yet!");
```
Comment (susilehtola) [inline psi4/src/psi4/adc/adc.cc:69]:
```suggestion
     if (nopen_) throw PSIEXCEPTION("Open-shell calculation has not been implemented yet!");
```

File: external/downstream/v2rdm_casscf/CMakeLists.txt (hunk 5 above)
Comment (JonathonMisiewicz) [reply external/downstream/v2rdm_casscf/CMakeLists.txt:19]: No, this is a change that I've discussed with Lori. This is an interim change for v2rdm tests not to fail.

Please ask for clarification before making demands.

File: psi4/src/psi4/libmints/wavefunction.cc (hunk 4 above)
Comment (JonathonMisiewicz) [reply psi4/src/psi4/libmints/wavefunction.cc:822]: I'm legitimately confused. `soccpi_` and `doccpi_` _are_ removed.

Like I said previously, the point of `docc` is to avoid recompute of `doccpi()` every iteration of the for loop.

File: psi4/src/psi4/adc/adc.cc (hunk 6 above)
Comment (JonathonMisiewicz) [reply psi4/src/psi4/adc/adc.cc:69]: Once again, I'll bother with this if I need to make other commits, but not otherwise. As ADC is not going to survive 1.7, this change isn't worth going through another test cycle.

File: psi4/src/psi4/libmints/wavefunction.cc (hunk 3 above)
Comment (susilehtola) [reply psi4/src/psi4/libmints/wavefunction.cc:759]: `bool sanity_check`?

File: psi4/src/psi4/libmints/wavefunction.cc [hunk 7]
```diff
@@ -784,18 +792,18 @@ void Wavefunction::set_reference_wavefunction(const std::shared_ptr<Wavefunction
...
-        doccpi_[h] = doccpi[h];
-        nalphapi_[h] = doccpi_[h] + soccpi_[h];
-        nbetapi_[h] = doccpi_[h];
+        nalphapi_[h] = doccpi[h] + socc[h];
+        nbetapi_[h] = doccpi[h];
     }
-    nalpha_ = doccpi_.sum() + soccpi_.sum();
-    nbeta_ = doccpi_.sum();
+    nalpha_ = nalphapi_.sum();
+    nbeta_ = nbetapi_.sum();
```
Comment (susilehtola) [inline psi4/src/psi4/libmints/wavefunction.cc:806]: Shouldn't there be a spin state check here?

File: psi4/src/psi4/libmints/wavefunction.cc [hunk 8]
```diff
@@ -784,18 +792,18 @@ void Wavefunction::set_reference_wavefunction(const std::shared_ptr<Wavefunction
 }
 
 void Wavefunction::force_doccpi(const Dimension &doccpi) {
```
Comment (susilehtola) [inline psi4/src/psi4/libmints/wavefunction.cc:794]:
```suggestion
void Wavefunction::force_doccpi(const Dimension & input_docc) {
```

File: psi4/src/psi4/libmints/wavefunction.cc [hunk 9]
```diff
@@ -784,18 +792,18 @@ void Wavefunction::set_reference_wavefunction(const std::shared_ptr<Wavefunction
...
+    auto socc = soccpi();
     for (int h = 0; h < nirrep_; h++) {
-        if ((soccpi_[h] + doccpi[h]) > nmopi_[h]) {
+        if ((socc[h] + doccpi[h]) > nmopi_[h]) {
```
Comment (susilehtola) [inline psi4/src/psi4/libmints/wavefunction.cc:797]:
```suggestion
        if ((socc[h] + input_docc[h]) > nmopi_[h]) {
```

File: psi4/src/psi4/libmints/wavefunction.cc [hunk 10]
```diff
@@ -784,18 +792,18 @@ void Wavefunction::set_reference_wavefunction(const std::shared_ptr<Wavefunction
...
             throw PSIEXCEPTION(
                 "Wavefunction::force_doccpi: Number of doubly and singly occupied orbitals in an irrep cannot exceed "
                 "the total number of molecular orbitals.");
         }
-        doccpi_[h] = doccpi[h];
-        nalphapi_[h] = doccpi_[h] + soccpi_[h];
-        nbetapi_[h] = doccpi_[h];
+        nalphapi_[h] = doccpi[h] + socc[h];
+        nbetapi_[h] = doccpi[h];
```
Comment (susilehtola) [inline psi4/src/psi4/libmints/wavefunction.cc:803]:
```suggestion
        nalphapi_[h] = input_docc[h] + socc[h];
        nbetapi_[h] = input_docc[h];
```

File: psi4/src/psi4/libmints/wavefunction.cc [hunk 11]
```diff
@@ -784,18 +792,18 @@ void Wavefunction::set_reference_wavefunction(const std::shared_ptr<Wavefunction
...
     }
-    nalpha_ = doccpi_.sum() + soccpi_.sum();
-    nbeta_ = doccpi_.sum();
+    nalpha_ = nalphapi_.sum();
+    nbeta_ = nbetapi_.sum();
 }
 
 void Wavefunction::force_soccpi(const Dimension &soccpi) {
```
Comment (susilehtola) [inline psi4/src/psi4/libmints/wavefunction.cc:809]:
```suggestion
void Wavefunction::force_soccpi(const Dimension & input_socc) {
```

File: psi4/src/psi4/libmints/wavefunction.cc [hunk 12]
```diff
@@ -804,16 +812,16 @@ void Wavefunction::force_soccpi(const Dimension &soccpi) {
             "Wavefunction::force_soccpi: Cannot set soccpi since alpha and beta densities must be the same for this "
             "Wavefunction.");
     }
+    auto docc = doccpi();
     for (int h = 0; h < nirrep_; h++) {
-        if ((soccpi[h] + doccpi_[h]) > nmopi_[h]) {
+        if ((soccpi[h] + doccpi()[h]) > nmopi_[h]) {
```
Comment (susilehtola) [inline psi4/src/psi4/libmints/wavefunction.cc:817]:
```suggestion
        if ((docc[h] + input_socc[h]) > nmopi_[h]) {
```

File: psi4/src/psi4/libmints/wavefunction.cc (hunk 4 above)
Comment (susilehtola) [reply psi4/src/psi4/libmints/wavefunction.cc:822]:
```suggestion
        nalphapi_[h] = docc[h] + input_socc[h];
```

File: psi4/src/psi4/libqt/qt.h [hunk 13]
```diff
@@ -59,10 +59,10 @@ void normalize(double** A, int rows, int cols);
 double invert_matrix(double** a, double** y, int N, std::string out_fname);
 void solve_2x2_pep(double** H, double S, double* evals, double** evecs);
 PSI_API
-void reorder_qt(int* docc_in, int* socc_in, int* frozen_docc_in, int* frozen_uocc_in, int* order, int* orbs_per_irrep,
+void reorder_qt(const int* docc_in, const int* socc_in, int* frozen_docc_in, int* frozen_uocc_in, int* order, int* orbs_per_irrep,
                 int nirreps);
 PSI_API
-void reorder_qt_uhf(int* docc, int* socc, int* frozen_docc, int* frozen_uocc, int* order_alpha, int* order_beta,
+void reorder_qt_uhf(const int* docc, const int* socc, int* frozen_docc, int* frozen_uocc, int* order_alpha, int* order_beta,
```
Comment (susilehtola) [inline psi4/src/psi4/libqt/qt.h:65]: These changes are unrelated to the present pull

File: psi4/src/psi4/libqt/reorder_qt.cc [hunk 14]
```diff
@@ -193,7 +193,7 @@ PSI_API void reorder_qt(int *docc_in, int *socc_in, int *frozen_docc_in, int *fr
 **
 ** \ingroup QT
 */
-PSI_API void reorder_qt_uhf(int *docc, int *socc, int *frozen_docc, int *frozen_uocc, int *order_alpha, int *order_beta,
+PSI_API void reorder_qt_uhf(const int *docc, const int *socc, int *frozen_docc, int *frozen_uocc, int *order_alpha, int *order_beta,
```
Comment (susilehtola) [inline psi4/src/psi4/libqt/reorder_qt.cc:196]: Same here

File: psi4/src/psi4/libmints/wavefunction.cc (hunk 3 above)
Comment (JonathonMisiewicz) [reply psi4/src/psi4/libmints/wavefunction.cc:759]: The name `sanity_check` is only appropriate if having more beta than alpha is _always_ inappropriate when this function is called. Lori and I talked among ourselves and thought it was best not to make that assumption. If that assumption _were_ correct, we should raise an error rather than warn, anyways.

File: psi4/src/psi4/libscf_solver/hf.cc [hunk 15]
```diff
@@ -208,15 +199,22 @@ void HF::common_init() {
         }
     }
 
+    // Check that we have enough basis functions
+    for (int h = 0; h < nirrep_; ++h) {
+        if (std::max(nalphapi_[h], nbetapi_[h]) > nsopi_[h]) {
+            throw PSIEXCEPTION("Not enough basis functions to satisfy requested occupancies");
```
Comment (susilehtola) [inline psi4/src/psi4/libscf_solver/hf.cc:205]:
```
             throw PSIEXCEPTION("Not enough linearly independent basis functions to satisfy requested occupancies");
```

File: psi4/src/psi4/libqt/qt.h (hunk 13 above)
Comment (JonathonMisiewicz) [reply psi4/src/psi4/libqt/qt.h:65]: Tests failed without this change - it's related.

Test failure was that the `const` return value of `doccpi()` or `soccpi()` was trying to be passed into this function.

File: psi4/src/psi4/libscf_solver/hf.cc [hunk 16]
```diff
@@ -1227,18 +1194,20 @@ void HF::check_phases() {
...
-    for (int h = 0; h < nirrep_ - 1; ++h) outfile->Printf(" %4d,", doccpi_[h]);
-    outfile->Printf(" %4d ]\n", doccpi_[nirrep_ - 1]);
+    for (int h = 0; h < nirrep_ - 1; ++h) outfile->Printf(" %4d,", docc[h]);
+    outfile->Printf(" %4d ]\n", docc[nirrep_ - 1]);
     if (reference != "RHF" && reference != "RKS") {
         outfile->Printf("    SOCC [ ");
-        for (int h = 0; h < nirrep_ - 1; ++h) outfile->Printf(" %4d,", soccpi_[h]);
-        outfile->Printf(" %4d ]\n", soccpi_[nirrep_ - 1]);
+        for (int h = 0; h < nirrep_ - 1; ++h) outfile->Printf(" %4d,", socc[h]);
+        outfile->Printf(" %4d ]\n", socc[nirrep_ - 1]);
```
Comment (susilehtola) [inline psi4/src/psi4/libscf_solver/hf.cc:1210]: The code is not printing out AOCC and BOCC

File: psi4/src/psi4/libqt/reorder_qt.cc (hunk 14 above)
Comment (JonathonMisiewicz) [reply psi4/src/psi4/libqt/reorder_qt.cc:196]: See discussion [here](https://github.com/psi4/psi4/pull/2619#discussion_r938033910).

File: psi4/src/psi4/libscf_solver/hf.cc (hunk 16 above)
Comment (JonathonMisiewicz) [reply psi4/src/psi4/libscf_solver/hf.cc:1210]: Correct. The _old behavior_ was to print out DOCC and SOCC. I didn't want to change it to print out AOCC and BOCC - too much for one PR.

File: psi4/src/psi4/libmints/wavefunction.cc (hunk 4 above)
Comment (JonathonMisiewicz) [reply psi4/src/psi4/libmints/wavefunction.cc:822]: You can add Dimension objects directly rather than doing it by irrep for further simplifications, but agreed that the function can and should be simplified now.

File: external/downstream/v2rdm_casscf/CMakeLists.txt (hunk 5 above)
Comment (susilehtola) [reply external/downstream/v2rdm_casscf/CMakeLists.txt:19]: > No, this is a change that I've discussed with Lori. This is an interim change for v2rdm tests not to fail.
> 
> Please ask for clarification before making demands.

See above: "What's the reason for this change? It does not appear to be related to docc/socc."

File: psi4/src/psi4/libmints/wavefunction.cc (hunk 7 above)
Comment (JonathonMisiewicz) [reply psi4/src/psi4/libmints/wavefunction.cc:806]: I don't know. The `force` functions are [marked](https://github.com/psi4/psi4/blob/master/psi4/src/psi4/libmints/wavefunction.h#L402-L416) as intended for expert use only because careless use can produce inconsistencies. Nothing in Psi uses them. I can't tell whether the potential to cause a spin state violation is a bug or a feature. Maybe a warning is in order if the number of unpaired electrons changes?

Comment (susilehtola) [reply psi4/src/psi4/libmints/wavefunction.cc:806]: OK, in that case we can forget about it

File: external/downstream/v2rdm_casscf/CMakeLists.txt (hunk 5 above)
Comment (JonathonMisiewicz) [reply external/downstream/v2rdm_casscf/CMakeLists.txt:19]: Current V2RDM uses `doccpi_` and `soccpi_`. This changes the copy of V2RDM we use for test purposes to a branch of Lori's where those have been replaced with the new function calls.

Comment (loriab) [reply external/downstream/v2rdm_casscf/CMakeLists.txt:19]: Because the wfn interface changed, the v2rdm plugin needed changes. there were some that arose from this PR and other from another one. last two commits of https://github.com/loriab/v2rdm_casscf/commits/v2rdm8 . the change above is loosening from a particular commit that was the branch head to the branch itself.

File: psi4/src/psi4/libmints/wavefunction.cc (hunk 7 above)
Comment (susilehtola) [reply psi4/src/psi4/libmints/wavefunction.cc:806]: Actually... I think the `force` functions may break with these changes, since `socc` and `docc` are recomputed. One would have to override `socc` and `docc` simultaneously lest the occupations be changed in the other code.

Comment (susilehtola) [reply psi4/src/psi4/libmints/wavefunction.cc:806]: I would remove the force functions altogether, since it sounds like they aren't even used..

File: psi4/src/psi4/libmints/wavefunction.cc (hunk 4 above)
Comment (susilehtola) [reply psi4/src/psi4/libmints/wavefunction.cc:822]: I was legitimately confused because the array names were inconsistent. The point of these functions is that an input array is passed in and then combined with the computed arrays. I have proposed above to use a more consistent naming scheme: `socc` and `docc` for the precomputed arrays and `input_socc` and `input_docc` for the overriding inputs.

File: psi4/src/psi4/libmints/wavefunction.cc (hunk 7 above)
Comment (JonathonMisiewicz) [reply psi4/src/psi4/libmints/wavefunction.cc:806]: @loriab Fine with you? These functions are dangerous. Forte doesn't use them. V2RDM doesn't use them? Psi doesn't use them.

Comment (loriab) [reply psi4/src/psi4/libmints/wavefunction.cc:806]: I need to look further. I think there was a Shannon Houck change that allowed resetting docc/socc for detci, and CDS agreed it was a legit use. Would have to see if this is that.

File: psi4/src/psi4/libscf_solver/hf.cc (hunk 16 above)
Comment (susilehtola) [reply psi4/src/psi4/libscf_solver/hf.cc:1210]: Does the code already print out the number of alpha and beta electrons? I don't think it does.

Since the code changes the way the calculation is run, the printout should be changed as well. Just add printouts for AOCC and BOCC?

Comment (JonathonMisiewicz) [reply psi4/src/psi4/libscf_solver/hf.cc:1210]: AOCC and BOCC are printed in the "excited MOM" case. I'm happy to change to print alpha and beta always, print docc and socc if well-defined. Again, fine with @loriab ?

Comment (loriab) [reply psi4/src/psi4/libscf_solver/hf.cc:1210]: more printing? sure!

File: psi4/src/psi4/libmints/wavefunction.cc (hunk 7 above)
Comment (loriab) [reply psi4/src/psi4/libmints/wavefunction.cc:806]: aha, https://github.com/psi4/psi4/pull/876 . the ability added in that PR should be preserved. it likely enables things in the fockCI plugin

Comment (JonathonMisiewicz) [reply psi4/src/psi4/libmints/wavefunction.cc:806]: I'll give that a look. Do we have FockCI test coverage?

Comment (JonathonMisiewicz) [reply psi4/src/psi4/libmints/wavefunction.cc:806]: * The express purpose of this forcing is to [change spin state](https://github.com/shannonhouck/psi4fockci/blob/267e17103c77dcd42cf1a66a5b805db3bcd5ed9e/psi4fockci/spinflip.py#L131-L142). So the lack of a spin state check is definitely a feature and not a bug.
* This use cases forces both `doccpi` and `soccpi`. It makes sense to me to have the user force both simultaneously.

Comment (susilehtola) [reply psi4/src/psi4/libmints/wavefunction.cc:806]: Then those functions definitely need to be merged into `force_docc_socc`; otherwise there is no guarantee that the code will actually use the input arrays.

Comment (loriab) [reply psi4/src/psi4/libmints/wavefunction.cc:806]: > I'll give that a look. Do we have FockCI test coverage?

possibly not, 😳 . She was working on two plugins, old and new. It was near a psi release, and I got very confused and packaged the wrong one. She added https://github.com/psi4/psi4/blob/master/doc/sphinxman/source/plugin_psi4fockci.rst though, so the internal build and that input might work.

Comment (loriab) [reply psi4/src/psi4/libmints/wavefunction.cc:806]: a combo docc_and_socc can work: https://github.com/shannonhouck/psi4fockci/commit/9a3916598030cdd501c7370ca2040bd7659e018b#diff-e33326fe2aaf9f7d7d385252ae30fa0c5a6d47751f890592c8ddc82bd8b861e4R83

Comment (JonathonMisiewicz) [reply psi4/src/psi4/libmints/wavefunction.cc:806]: > > I'll give that a look. Do we have FockCI test coverage?
> 
> possibly not, 😳 . She was working on two plugins, old and new. It was near a psi release, and I got very confused and packaged the wrong one. She added https://github.com/psi4/psi4/blob/master/doc/sphinxman/source/plugin_psi4fockci.rst though, so the internal build and that input might work.

What does "internal build" mean?

Comment (loriab) [reply psi4/src/psi4/libmints/wavefunction.cc:806]: toggle `ENABLE_psi4fockci=ON` in `<objdir>/CMakeCache.txt` and rebuild. hopefully cmake will pull a copy and install it into `stage/` so it's ready to run.

File: psi4/src/psi4/libmints/wavefunction.cc (hunk 3 above)
Comment (JonathonMisiewicz) [reply psi4/src/psi4/libmints/wavefunction.cc:759]: Renaming to `warn_on_beta_socc`, given aforementioned objections to `sanity_check`. Happy to re-discuss if there's a new alternative.

**Comment by JonathonMisiewicz** [comment]
FYI, I'm in the process of revising the PR and am marking comment as resolved as I fix them locally.

File: psi4/src/psi4/libscf_solver/hf.cc (hunk 15 above)
Comment (JonathonMisiewicz) [reply psi4/src/psi4/libscf_solver/hf.cc:205]: Proposed rewording is misleading. This exception is thrown due to a comparison against `nsopi`, which is the number of basis functions _regardless of_ linear dependence checking. (Linear dependence checking hasn't even occurred at the time this method is called.)

I suspect but have not confirmed that a system with enough basis functions, but not enough linearly independent basis functions, would escape this check and crash Psi. If true, we should file an issue.

File: psi4/src/psi4/libscf_solver/hf.cc (hunk 16 above)
Comment (JonathonMisiewicz) [reply psi4/src/psi4/libscf_solver/hf.cc:1210]: Changed proposal to "print alpha and beta always, print docc and socc **always**." A warning will fire if those might be misleading, and AOCC and BOCC should clear confusion.

File: psi4/src/psi4/libmints/wavefunction.cc (hunk 7 above)
Comment (JonathonMisiewicz) [reply psi4/src/psi4/libmints/wavefunction.cc:798]: @loriab It doesn't look like FockCI has any tests registered with Psi. Am I mistaken?

**Comment by JonathonMisiewicz** [comment]
All comments except FockCI testing addressed.

File: psi4/src/psi4/libscf_solver/hf.cc (hunk 15 above)
Comment (susilehtola) [reply psi4/src/psi4/libscf_solver/hf.cc:205]: OK thanks for the clarification; I assumed that the number of symmetry functions was done *after* the linear dependence check.

File: psi4/src/psi4/libmints/wavefunction.h [hunk 17]
```diff
@@ -375,9 +375,9 @@ class PSI_API Wavefunction : public std::enable_shared_from_this<Wavefunction> {
     static void initialize_singletons();
 
     /// Returns the DOCC per irrep array. Not recommended for unrestricted code.
-    const Dimension doccpi(bool assume_socc_alpha = true) const;
+    const Dimension doccpi(bool warn_on_beta_socc = true) const;
     /// Returns the SOCC per irrep array. Not recommended for unrestricted code.
```
Comment (susilehtola) [inline psi4/src/psi4/libmints/wavefunction.h:379]: Shouldn't the argument be documented here? e.g. `warn_on_beta_socc: trigger warning on singly occupied beta orbitals that break assumptions made in the old-fashioned DOCC/SOCC handling`

File: psi4/src/psi4/libmints/wavefunction.cc (hunk 7 above)
Comment (loriab) [reply psi4/src/psi4/libmints/wavefunction.cc:798]: you could try https://github.com/psi4/psi4/blob/master/tests/pytests/test_addons.py#L1236 . because of my repo confusion, though, it may not be what you're after.

**Comment by JonathonMisiewicz** [comment]
@susilehtola Comments added.

Per discussion with @loriab , I'll leave a comment on the FockCI repo letting them know about this API-breaking change. It sounds like that repo has its own issues that need to be dealt with, beyond the scope of this PR.

File: psi4/src/psi4/adc/adc.cc [hunk 18]
```diff
@@ -45,53 +45,48 @@ ADCWfn::ADCWfn(SharedWavefunction ref_wfn, Options& options) : Wavefunction(opti
...
     }
 
+    const auto docc = doccpi();
+    const auto socc = soccpi();
     outfile->Printf("\n\n\tIrrep  Core  Docc  Socc  aOcc  aVir  bOcc  bVir  FVir\n");
     outfile->Printf("\t*****************************************************\n");
     for (int h = 0; h < nirrep_; h++) {
         outfile->Printf("\t %3s   %3d   %3d   %3d   %3d   %3d   %3d   %3d   %3d\n", irreps_[h].c_str(), frzcpi_[h],
-                        doccpi_[h], soccpi_[h], aoccpi_[h], avirpi_[h], boccpi_[h], bvirpi_[h], frzvpi_[h]);
+                        docc[h] + socc[h], soccpi()[h], aoccpi_[h], avirpi_[h], boccpi_[h], bvirpi_[h], frzvpi_[h]);
```
Comment (susilehtola) [inline psi4/src/psi4/adc/adc.cc:89]: You're supposed to print docc and socc so
```suggestion
                        docc[h], socc[h], aoccpi_[h], avirpi_[h], boccpi_[h], bvirpi_[h], frzvpi_[h]);
```

File: psi4/src/psi4/libmints/wavefunction.cc [hunk 19]
```diff
@@ -753,6 +739,28 @@ void Wavefunction::initialize_singletons() {
     done = true;
 }
 
+const Dimension Wavefunction::doccpi(bool warn_on_beta_socc) const {
+    std::vector<int> docc_vec;
+    for (int h = 0; h < nalphapi_.n(); h++) {
+        docc_vec.push_back(std::min(nalphapi_[h], nbetapi_[h]));
+        if (warn_on_beta_socc && nbetapi_[h] > nalphapi_[h]) {
+            outfile->Printf("Warning! Irrep has more beta than alpha electrons.\n");
```
Comment (susilehtola) [inline psi4/src/psi4/libmints/wavefunction.cc:747]: .. in symmetry `h`

File: psi4/src/psi4/libmints/wavefunction.cc [hunk 20]
```diff
@@ -753,6 +739,28 @@ void Wavefunction::initialize_singletons() {
...
+    }
+    return docc_vec;
+}
+/// Returns the SOCC per irrep array. Not recommended for unrestricted code.
+const Dimension Wavefunction::soccpi(bool warn_on_beta_socc) const {
+    std::vector<int> socc_vec;
+    for (int h = 0; h < nalphapi_.n(); h++) {
+        socc_vec.push_back(std::abs(nalphapi_[h] - nbetapi_[h]));
+        if (warn_on_beta_socc && nbetapi_[h] > nalphapi_[h]) {
+            outfile->Printf("Warning! Irrep has more beta than alpha electrons.\n");
```
Comment (susilehtola) [inline psi4/src/psi4/libmints/wavefunction.cc:758]: same here

**Comment by susilehtola** [comment]
@JonathonMisiewicz as far as I can see, this should be OK now except the one wrong printout above.

**Review by susilehtola — APPROVED** [review]

File: psi4/src/psi4/libmints/wavefunction.cc (hunk 19 above)
Comment (JonathonMisiewicz) [reply psi4/src/psi4/libmints/wavefunction.cc]: Changed to "in symmetry `h` orbitals" instead, but done in principle.

**Review by zachglick — APPROVED** [review]
LGTM, thanks for working on this. Any reason for not adding a test or two?

**Comment by JonathonMisiewicz** [comment]
> LGTM, thanks for working on this. Any reason for not adding a test or two?

No tests came to mind as worth adding. If you have suggestions, I'll probably take them.

**Comment by zachglick** [comment]
No suggestions, merge when ready.

**Comment by JonathonMisiewicz** [comment]
[PRd to FockCI](https://github.com/shannonhouck/psi4fockci/pull/7), so I can wash my hands of the `force_.occpi` changes.

**Comment by susilehtola** [comment]
> > LGTM, thanks for working on this. Any reason for not adding a test or two?
> 
> No tests came to mind as worth adding. If you have suggestions, I'll probably take them.

Any one of the various failing SCF calculations where occupations break old invalid assumptions?

**Comment by JonathonMisiewicz** [comment]
> A subsequent PR will fix the linked issues (if not fixed by this PR) and add them as test cases once the fix is confirmed.

On my to-do list, I just wasn't confident those would already be fixed by this PR. Thanks for the reminder.
